File: app/src/main/kotlin/com/openclaw/ghostcrab/domain/repository/DiscoveryService.kt
package com.openclaw.ghostcrab.domain.repository

import com.openclaw.ghostcrab.domain.model.DiscoveredGateway
import kotlinx.coroutines.flow.Flow

/**
 * Discovers OpenClaw Gateway instances on the local network via mDNS (NsdManager).
 *
 * Service type: `_openclaw-gw._tcp.` Default port: 18789.
 *
 * **Contract frozen at v1.0.**
 */
interface DiscoveryService {

    /**
     * Starts mDNS discovery and emits each resolved gateway as it is found.
     *
     * - Acquires a `WifiManager.MulticastLock` before starting.
     * - Deduplicates by `instanceName`; subsequent resolutions of the same instance are ignored.
     * - The flow completes when [stopDiscovery] is called or after the caller's collection scope
     *   is cancelled.
     *
     * @return A cold [Flow] that emits [DiscoveredGateway] instances as they are resolved.
     *   Does not throw; resolution failures for individual services are silently skipped.
     */
    fun startDiscovery(): Flow<DiscoveredGateway>

    /**
     * Stops any active mDNS discovery and releases the multicast lock.
     *
     * Safe to call when discovery is not active.
     */
    suspend fun stopDiscovery()
}

File: ios/GhostCrab/Domain/Repository/DiscoveryService.swift
import Foundation

/// Discovers OpenClaw Gateway instances on the local network via mDNS (`NWBrowser`).
///
/// Service type: `_openclaw-gw._tcp.` Default port: 18789.
///
/// **Contract frozen at v1.0.**
///
/// > Important: The host app's `Info.plist` must declare `NSBonjourServices`
/// > containing `_openclaw-gw._tcp` or iOS silently blocks the browser.
public protocol DiscoveryService: Sendable {

    /// Starts mDNS discovery and emits each resolved gateway as it is found.
    ///
    /// - Acquires the local-network entitlement implicitly via `NWBrowser`.
    /// - Deduplicates by `instanceName`; subsequent resolutions of the same instance
    ///   are ignored.
    /// - The stream completes when `stopDiscovery()` is called or the consuming task
    ///   is cancelled.
    ///
    /// - Returns: An `AsyncStream` that emits `DiscoveredGateway` instances as they
    ///   are resolved. Does not throw; resolution failures for individual services are
    ///   silently skipped.
    func startDiscovery() -> AsyncStream<DiscoveredGateway>

    /// Stops any active mDNS discovery.
    ///
    /// Safe to call when discovery is not active.
    func stopDiscovery() async
}
